paimon-core/src/main/java/org/apache/paimon/manifest/ManifestFileMeta.java

```diff
@@ -185,4 +207,108 @@ private static void mergeCandidates(
+                    BiFunction<Predicate, List<ManifestFileMeta>, Integer>
+                            firstValidManifestFunction =
+                                    (predicate, manifests) -> {
+                                        FieldStatsArraySerializer fieldStatsArraySerializer =
+                                                new FieldStatsArraySerializer(partitionType);
+                                        int firstMatchIndex = manifests.size();
+                                        for (int i = 0; i < manifests.size(); i++) {
+                                            ManifestFileMeta manifestFileMeta = manifests.get(i);
+                                            if (paritionPredicate.test(
+                                                    manifestFileMeta.numAddedFiles
+                                                            + manifestFileMeta.numDeletedFiles,
+                                                    manifestFileMeta
+                                                            .partitionStats()
+                                                            .fields(fieldStatsArraySerializer))) {
+                                                firstMatchIndex = i;
+                                                break;
+                                            }
+                                        }
+                                        return firstMatchIndex;
+                                    };
+
+                    index = firstValidManifestFunction.apply(paritionPredicate, metas);
+                } else {
+                    index = 0;
+                }
+            }
+
+            return index;
+        }
+
+        public boolean needFullCompaction() {
+            return index < metas.size();
+        }
+
+        private Predicate convertPartitionToPredicate(
+                RowType partitionType, Set<BinaryRow> partitions) {
+            RowDataToObjectArrayConverter rowDataToObjectArrayConverter =
+                    new RowDataToObjectArrayConverter(partitionType);
+
+            List<Predicate> predicateList =
+                    partitions.stream()
+                            .map(
+                                    partition ->
+                                            AbstractFileStoreScan.partitionToPredicate(
+                                                    partition, rowDataToObjectArrayConverter))
+                            .collect(Collectors.toList());
+            Predicate paritionPredicate = PredicateBuilder.or(predicateList);
+            return paritionPredicate;
+        }
+
+        private Set<BinaryRow> findPartitionsToDeleteData(
+                ManifestFile manifestFile, List<ManifestFileMeta> manifestContainDelete) {
+            Set<BinaryRow> partitions = new HashSet<>();
+            for (ManifestFileMeta meta : manifestContainDelete) {
+                List<ManifestEntry> manifestEntries = manifestFile.read(meta.fileName);
```

Comment:
Maybe we can just do a merge here. Do merge and collect partitions for delete entries...